### ISSUE TITLE
Refactor `prove_equality_from_rule` in `execution_proof_generation.py`

### ISSUE THREAD

**Issue by fiedlr** [runtimeverification/proof-checker#629]
The solution in https://github.com/runtimeverification/proof-checker/issues/601 that was implemented is a bit adhoc and does not deal with all the edge cases.

This is a follow-up that refactors `prove_equality_from_rule` based on the systematic approach described in https://github.com/runtimeverification/proof-checker/pull/620

We need to remove all the added lemmas and replace them with these ones:

```
T /\ phi = phi
phi /\ T = phi
phi = psi -> psi = phi
phi <= phi -> Top
phi /\ phi = phi

psi1 = psi2  
------------------------------------
phi1[psi1/x]  = phi1[psi2/x]
for any variable `x

psi1 = psi2   phi1[psi1/x] 
------------------------------------
phi1[psi2/x]
for any variable `x
``` 

The one before last is an equivalent version of the equality congruence proved in the matching mu-logic TR, and a stronger version of equality_with_subst.

The last one is a semantic consequence of the previous where both sides are Top. It is also a stronger version of `lift_through_rewrite`

**Comment by IljaZakharov** [comment]
We need an extra lemma for the implication, like this:

```
kT -> phi
---------
    phi
```
Check out the proof:

![Screenshot 2023-12-14 at 10 56 16](https://github.com/runtimeverification/proof-checker/assets/1668230/2d4db076-f07d-4ccf-a6ed-b8022f6c3d84)

**Comment by fiedlr** [comment]
That is correct, @IljaZakharov , the task supposes this derived rule is not removed from the current file and left as is.
